Roll mockito to 5.4.0

natebosch opened the pull request:
The pinned version of `mockito` is incompatible with the latest released version of `test_api`, and therefore `test`. With flutter package pinning, the release of the flutter SDK that updates to the next version of `test_api` will break the version solve with these packages. Use the `5.4.0` version of `mockito` which is forwards compatible.

This is required to unblock https://github.com/flutter/flutter/pull/123350, https://github.com/flutter/flutter/issues/123603

stuartmorgan commented:
Mockito is pinned deliberately (which is why this is failing repo checks); this regresses https://github.com/flutter/packages/pull/3466. See also https://github.com/flutter/flutter/issues/108555, which is an OOB breakage affecting multiple trees that was caused by not pinning mockito.

What version of mockito is compatible with the update? Can we just roll the pin?

natebosch commented:
Mockito version `5.4.0` will be compatible.

It is currently not published because it was retracted by @zanderso , but when it is published it should be possible to upgrade to it.

> Mockito is pinned deliberately

That's a semver anti-pattern and you should expect to not be cross-compatible with different versions of `flutter_test` in that some cases. We will need to find some way to commit changes like https://github.com/flutter/flutter/pull/123350 with a red CI and then update the pin or something similar if we insist on pinning.

Package pinning is, in my opinion, an anti-solution to the problem which is misusing deprecated diagnostics.

stuartmorgan commented:
> We will need to find some way to commit changes like [flutter/flutter#123350](https://github.com/flutter/flutter/pull/123350) with a red CI and then update the pin or something similar if we insist on pinning.

I don't see why we couldn't just update flutter/packages with the new version of mockito first, then roll flutter/flutter, without anything having to be landed on red.

> which is that we are misusing deprecated diagnostics.

How so?

natebosch commented:
FWIW I think we misuse deprecations in other places too - flutter isn't the only one - but @lrhn and do get blocked from informing our users early about deprecations because of "strict" CI like this.

stuartmorgan commented:
> The reason to deprecate an API instead of removing it is so that there is a release which is _non-breaking_ for consumers and gives them a heads up.

I understand the concept of deprecation periods. What is the mechanism by which you propose that the Flutter team *receive* that heads up without surfacing it in CI?

> Treating the the release as if it was _breaking_ instead eliminates the value of the deprecation.

You are overloading the term "breaking" here though. There's a vast difference between breaking our CI and breaking clients of our package. Not breaking clients has enormous value.

> > I don't see why we couldn't just update flutter/packages with the new version of mockito first, then roll flutter/flutter, without anything having to be landed on red.
> 
> There is no guarantee that we can make `mockito` forwards compatible in every case. It should work in this case.

We can adjust tooling to allow non-carat ranges instead of only exact pins, such that we can set the upper range to a known-working, already-released version.

natebosch commented:
> You are overloading the term "breaking" here though.

I'm using it the way that I see it when this issue surfaces.

"This is breaking the flutter SDK, you must roll this back".

> I understand the concept of deprecation periods. What is the mechanism by which you propose that the Flutter team _receive_ that heads up without surfacing it in CI?

We need some warning level that provides feedback but isn't an error. I don't know the best way to surface that through our various CI.

natebosch commented:
> We can adjust tooling to allow non-carat ranges instead of only exact pins, such that we can set the upper range to a known-working, already-released version.

I think that should work. It makes the critical path longer, and it increases the general friction around flutter's pinned packages.

stuartmorgan commented:
> We need some warning level that provides feedback but isn't an error. I don't know the best way to surface that through our various CI.

I'm not aware of a mechanism to do that.

> It makes the critical path longer, and it increases the general friction around flutter's pinned packages.

Compared to what? The other options in aware of that are available to us are:
- Don't pin, keep deprecation warnings. This leads to OOB mutli-tree breakage events like the one I liked to, which are very high friction.
- Don't pin, disable deprecation warnings. This means we won't find out about deprecation warnings in *any package we use*. I'm skeptical that never finding out about deprecation, until we need to upgrade for some reason and suddenly discover a backlog of deprecated API use that we must immediately fix, will decrease friction overall. (I assume given your comments above that you would at least partially agree, because avoiding that is the whole point of deprecation warnings.)

I'm open to suggestions if you have them. I didn't choose the current approach for this repo because I think it's ideal, but because it seemed like the best available option.

natebosch commented:
> Compared to what?

Compared to not needing a PR in `flutter/packages` before we can roll the test packages in `flutter/flutter`.

stuartmorgan commented:
> > Compared to what?
> 
> Compared to not needing a PR in `flutter/packages` before we can roll the test packages in `flutter/flutter`.

That's an outcome, not an approach. I've explained the approaches that I am aware of for achieving that specific outcome, and the corresponding friction they create in other ways.

natebosch commented on packages/camera/camera_android_camerax/pubspec.yaml:
```diff
@@ -31,4 +31,4 @@ dev_dependencies:
     sdk: flutter
   mockito: 5.4.0
-  pigeon: ^3.2.6
+  pigeon: ^9.1.0
```
@stuartmorgan - I cannot find any file specifying the arguments to `flutter pub run pigeon` to use for this package, so I cannot test that the breaking changes in `pigeon` in the last 8 months do not impact this package.

How can I find the command to run?

stuartmorgan replied:
It's probably just `--input pigeons/camerax_library.dart`, but I don't work on that package so I don't know; @camsim99 would be the best person to answer that. (Tooling for keeping Pigeon updated in the repo is on my TODO list, FWIW.)

Any Pigeon updates should be done as separate precursor PRs since they impact production code.

gmackall replied:
`flutter pub run pigeon --input pigeons/camerax_library.dart` is what I use when working on this package

natebosch replied:
> Any Pigeon updates should be done as separate precursor PRs since they impact production code.

https://github.com/flutter/packages/pull/3561

Does this package impact google3? Or does it impact production without being published?

stuartmorgan replied:
> Or does it impact production without being published?

I'm not sure I understand the question. The generated code is production code, which means once it's published it affects clients. And since it affects clients, it needs to be published by repo policy.

natebosch replied:
> I'm not sure I understand the question.

I was thrown off by the `publish_to: none`, but I guess one of the files is generated into some other package that is published?

https://github.com/flutter/packages/blob/e893fdc9617c84b69737edbbdc681501289ed8d4/packages/camera/camera_android_camerax/pubspec.yaml#L5

natebosch commented:
@stuartmorgan - these changes only impact local development. From my reading of https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates this change should not change versions or add `## NEXT` entries.

stuartmorgan commented:
That's correct for the changes so far. It looks like mockito 5.4.0 requires Flutter 3.3 though, which will require SDK version bumps, and those will need `NEXT` entries. (I can take care of that for you with our repo tools right now.)

stuartmorgan commented:
Overriding version: mockito changes are dev-only, and SDK changes we deliberately don't do releases for since they are purely limiting.

ditman reviewed (approved):
Web packages LGTM!, thanks for the update!

(Note that this rolls mockito to 5.4.0 but it doesn't seem to update the mock files, so we should expect a wave of new mock files as people contribute to the affected packages (HOPEFULLY :P))

natebosch commented:
> doesn't seem to update the mock files

I found some `extends Mock` and then stopped looking -  I should have checked more directories. I've regenerated the mocks in the packages that were using codegen.

ditman commented:
I'm missing some updates of .mock.dart I'd expect to see, and I think it's because they're in `example` apps that are used to run `integration_test`s of those packages:

```console
$ grep "mockito: 5.3.2" * -R | grep example
google_maps_flutter/google_maps_flutter_web/example/pubspec.yaml:  mockito: 5.3.2
google_sign_in/google_sign_in_web/example/pubspec.yaml:  mockito: 5.3.2
url_launcher/url_launcher_ios/example/pubspec.yaml:  mockito: 5.3.2
url_launcher/url_launcher_android/example/pubspec.yaml:  mockito: 5.3.2
url_launcher/url_launcher/example/pubspec.yaml:  mockito: 5.3.2
url_launcher/url_launcher_web/example/pubspec.yaml:  mockito: 5.3.2
```

I can update those in this branch, if you want @natebosch.

natebosch commented:
> and they stay the same,

Hmm, I would expect at least the `// Mocks generated by Mockito 5.3.2` to change...

natebosch commented:
I added some `generate_for` config and regenerated mocks in packages using a directory named `integraion_test/` instead of `test/`. I also removed a stale mock file that was no longer used at all.

natebosch commented:
https://cirrus-ci.com/task/5467466374053888

```
The following packages had errors:
  camera_android_camerax
See above for full details.
```

Searching the page for `camera_android_camerax` does not find any other occurrences...

Is there some other way to see the logs for what failed?

ditman commented:
`submit-queue` is just the check that stops merges to `main` if there's currently a broken test in the top of the repo. This PR looks ready to land.

Thanks for digging in the unchanged mocks! @natebosch

stuartmorgan reviewed (approved):
Final version LGTM as well. Thanks!
